QUIC APIs (#4)
CarnaViire/blogpost-7.0: quic-api.md

# QUIC

> QUIC is a new, transport layer protocol. It has been recently standardized in [RFC 9000](https://www.rfc-editor.org/rfc/rfc9000.html). It uses UDP as an underlying protocol and it's inherently secure as it mandates TLS 1.3 usage, see [RFC 9001](https://www.rfc-editor.org/rfc/rfc9001.html). Another interesting difference from well known transport protocols as TCP and UDP is that it offers stream multiplexing over a shared connection. The connection between two endpoints negotiates secrets and settings and then the individual streams exchange data between the peers. This allows to have multiple, concurrent, detached data streams that do not affect each other.

Reviewer: > This allows to have multiple, concurrent, detached data streams that do not affect each other.

It might be only me, but "detached" sounds more like a stream is detached from a connection... Maybe "independent" will sound better

Author: Done

QUIC itself doesn't define any semantics for the exchanged data as it's a transport protocol but it's used in application layer protocols. For example in [HTTP/3](https://www.rfc-editor.org/rfc/rfc9114.html) or in [SMB over QUIC](https://learn.microsoft.com/en-us/windows-server/storage/file-server/smb-over-quic). It can also be used for any custom defined protocol.

> The protocol offers many advantages over TCP with TLS. For example, faster connection establishment as it doesn't require as many round trips as TCP with TLS on top. Or avoidance of head-of-line blocking problem where one lost datagram doesn't block data of all the other streams. On the other hand, there are disadvantages that come with using QUIC. As it is a new protocol, its adoption is still growing and is limited. Apart from that, QUIC traffic might be even blocked by some networking components.

Reviewer: > Or avoidance of head-of-line blocking problem where one lost datagram doesn't block data of all the other streams.

Should we use term "packet" instead of "datagram" here?

Author: Done.

## QUIC in .NET

> We introduced QUIC implementation in .NET 5 in `System.Net.Quic` library. However up until now, the library was strictly internal and served only for own implementation of HTTP/3. With the release of .NET 7, we're making the library public and we're exposing its APIs. Since we had only `HttpClient` and Kestrel as consumers of the APIs, we decided to keep them as [preview feature](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md). It gives us ability to tweak the API in the next release before we settle on the final shape.

Reviewer: > Since we had only `HttpClient` and Kestrel as consumers of the APIs, we decided to keep them as [preview feature](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md).

Maybe we should add "for this release".

Author: Done.

> From the implementation perspective, `System.Net.Quic` depends on [MsQuic](https://github.com/microsoft/msquic), native implementation of the QUIC protocol. As a result, `System.Net.Quic` platform support and dependencies are inherited from MsQuic and documented in [HTTP/3 Platform dependencies](https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-http3#platform-dependencies). It is still possible to build MsQuic manually, whether against SChannel or OpenSSL, and use it with `System.Net.Quic`. However, these scenarios are not part of our testing matrix and unforeseen problems might occur.

Reviewer: I think we should add something like "note that on Linux you should [manually](https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-http3#linux) install MsQuic (`libmsquic`) on your machine", -- because later we say "`libmsquic` might be missing" but it is not clear enough what we are referring to and that it is specific to Linux.

Author: Going with:  "In short, MsQuic library is shipped as part of .NET for Windows, but for Linux libmsquic must be manually installed via a package manager. "

The sample usage of `QuicListener`:
```C#
using System.Net.Quic;

// First, check if QUIC is supported.
if (!QuicListener.IsSupported)
{
    Console.WriteLine("QUIC is not supported, check for presence of libmsquic and support of TLS 1.3.");
    return;
}

// We want the same configuration for each incoming connection, so we prepare the connection option upfront and reuse them.
// This represents the minimal configuration necessary to accept a connection.
var serverConnectionOptions = new QuicServerConnectionOptions()
{

>     // Used to abort stream if it's not properly closed by the user.

Reviewer: Maybe we should add something like "Can be any number as defined by the application level protocol"

Author: Instead of that, I'm going with the real H/3 error codes as I'm using "h3" ALPN. I think that concrete example is here better.

Author: Used different constant and added "Protocol-dependent error code." comment

    // See https://www.rfc-editor.org/rfc/rfc9000.html#name-application-protocol-error-
    DefaultStreamErrorCode = 123,

    // Used to close the connection if it's not done by the user.
    // See https://www.rfc-editor.org/rfc/rfc9000.html#name-application-protocol-error-
    DefaultCloseErrorCode = 456,

    // Same options as for server side SslStream.
    ServerAuthenticationOptions = new SslServerAuthenticationOptions
    {
        // List of supported application protocols, must be the same or subset of QuicListenerOptions.ApplicationProtocols.

>         ApplicationProtocols = new List<SslApplicationProtocol>() { SslApplicationProtocol.Http3 },

Reviewer: I think we should use a different ALPN for example purposes, e.g. "quic-test"

Author: Aaaa, going another way and using other H/3 stuff for our samples. Personally, I prefer concrete, existing examples.

Author: Used a string constant.

# QUIC

> QUIC is a new, transport layer protocol. It has been recently standardized in [RFC 9000](https://www.rfc-editor.org/rfc/rfc9000.html). It uses UDP as an underlying protocol and it's inherently secure as it mandates TLS 1.3 usage, see [RFC 9001](https://www.rfc-editor.org/rfc/rfc9001.html). Another interesting difference from well known transport protocols as TCP and UDP is that it offers stream multiplexing over a shared connection. The connection between two endpoints negotiates secrets and settings and then the individual streams exchange data between the peers. This allows to have multiple, concurrent, independent data streams that do not affect each other.

Reviewer 2: > The connection between two endpoints negotiates secrets and settings and then the individual streams exchange data between the peers.

This sentence sounds rather weird to me and out of place in the paragraph

```suggestion
QUIC is a new, transport layer protocol. It has been recently standardized in [RFC 9000](https://www.rfc-editor.org/rfc/rfc9000.html). It uses UDP as an underlying protocol and it's inherently secure as it mandates TLS 1.3 usage, see [RFC 9001](https://www.rfc-editor.org/rfc/rfc9001.html). Another interesting difference from well-known transport protocols such as TCP and UDP is that it has stream multiplexing built-in on the transport layer.  This allows having multiple, concurrent, independent data streams that do not affect each other.
```

Author: Done.

        ApplicationProtocols = new List<SslApplicationProtocol>() { SslApplicationProtocol.Http3 }
    }
};

// Initialize, configure and connect to the server.
var connection = await QuicConnection.ConnectAsync(clientConnectionOptions);

Console.WriteLine($"Connected {connection.LocalEndPoint} --> {connection.RemoteEndPoint}");

// Open a bidirectional (can both read and write) outbound stream.
var outgoingStream = await connection.OpenOutboundStreamAsync(QuicStreamType.Bidirectional);

// Work with the outgoing stream ...

// To accept any stream on a client connection, at least one of MaxInboundBidirectionalStreams or MaxInboundUnidirectionalStreams of QuicConnectionOptions must be set.
while (isRunning)
{
    // Accept an inbound stream.
    var incomingStream = await connection.AcceptInboundStreamAsync();

    // Work with the incoming stream ...
}

// Close the connection with the custom code.
await connection.CloseAsync(789);

// Dispose the connection.
await connection.DisposeAsync();
```

More details about how this class was designed can be found in the [`QuicConnection` API Proposal](https://github.com/dotnet/runtime/issues/68902) issue.

### QuicStream

> [`QuicStream`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quicstream?view=net-7.0) is the actual type that is used to send and receive data in QUIC protocol. It derives from ordinary [`Stream`](https://learn.microsoft.com/en-us/dotnet/api/system.io.stream?view=net-7.0) and can be used as such. But it also offers several features that are specific to QUIC protocol. Firstly, a QUIC stream can either be unidirectional or bidirectional, see [RFC 9000 - 2.1. Stream Types and Identifiers](https://www.rfc-editor.org/rfc/rfc9000.html#name-stream-types-and-identifier). A bidirectional stream is able to send and receive data on both sides, whereas unidirectional stream can only write from the initiating side and read on the accepting one. A connection can limit how many concurrent stream of each type is willing to accept, see [`QuicConnectionOptions.MaxInboundBidirectionalStreams`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quicconnectionoptions.maxinboundbidirectionalstreams?view=net-7.0) and [`QuicConnectionOptions.MaxInboundUnidirectionalStreams`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quicconnectionoptions.maxinboundunidirectionalstreams?view=net-7.0).

Reviewer 2: ```suggestion
[`QuicStream`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quicstream?view=net-7.0) is the actual type that is used to send and receive data in QUIC protocol. It derives from ordinary [`Stream`](https://learn.microsoft.com/en-us/dotnet/api/system.io.stream?view=net-7.0) and can be used as such. But it also offers several features that are specific to QUIC protocol. Firstly, a QUIC stream can either be unidirectional or bidirectional, see [RFC 9000 - 2.1. Stream Types and Identifiers](https://www.rfc-editor.org/rfc/rfc9000.html#name-stream-types-and-identifier). A bidirectional stream is able to send and receive data on both sides, whereas a unidirectional stream can only write from the initiating side and read on the accepting one. Each peer can limit how many concurrent streams of each type it is willing to accept, see [`QuicConnectionOptions.MaxInboundBidirectionalStreams`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quicconnectionoptions.maxinboundbidirectionalstreams?view=net-7.0) and [`QuicConnectionOptions.MaxInboundUnidirectionalStreams`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quicconnectionoptions.maxinboundunidirectionalstreams?view=net-7.0).
```

Author: Done

|            | client  | server  |
| -          | -       | -       |

> | `CanRead`  | _bidirectional_: `true`<br/> _unidirectional_: `false` | `true`  |

Reviewer: Table is cool!
The only thing: the way it is written now, you can think that only client can open streams to server, which is not true...

Author: I'm adding a note before the table and also renaming the headers to "peer opening/accepting stream"

| `CanWrite` | `true`  | _bidirectional_: `true`<br/> _unidirectional_: `false` |
| `ReadAsync` | _bidirectional_: reads data<br/> _unidirectional_: `InvalidOperationException` | reads data |
| `WriteAsync` | sends data => server read returns the data | _bidirectional_: sends data => client read returns the data<br/> _unidirectional_: `InvalidOperationException`  |
| `CompleteWrites` | closes writing side => server read returns 0 | _bidirectional_: closes writing side => client read returns 0<br/> _unidirectional_: - |
| `Abort(QuicAbortDirection.Read)` | _bidirectional_: [STOP_SENDING](https://www.rfc-editor.org/rfc/rfc9000.html#name-stop_sending-frames) => server write throws `QuicException(QuicError.OperationAborted)`<br/> _unidirectional_: - | [STOP_SENDING](https://www.rfc-editor.org/rfc/rfc9000.html#name-stop_sending-frames) => client write throws `QuicException(QuicError.OperationAborted)`|

> | `Abort(QuicAbortDirection.Write)` | [RESET_STREAM](https://www.rfc-editor.org/rfc/rfc9000.html#name-reset_stream-frames) => server read throws `QuicException(QuicError.OperationAborted)` | _bidirectional_: [RESET_STREAM](https://www.rfc-editor.org/rfc/rfc9000.html#name-reset_stream-frames) => client read throws `QuicException(QuicError.OperationAborted)`<br/> _unidirectional_: - |

Reviewer: just my thoughts, would it be more clear, if instead of "-" in

> _unidirectional_: -

we would put something like "no-op"?

Author: Done.
